hibernate-rx-core/src/test/java/org/hibernate/rx/CascadeComplicatedTest.java

```diff
@@ -0,0 +1,442 @@
+ *     |         |                      |
+ *     |         |                      |
+ *     1         N                      |
+ *     *                                |
+ *     B * N---1 D * 1------------------
+ *     *
+ *     N         N
+ *     |         |
+ *     |         |
+ *     1         |
+ *               |
+ *     C * 1-----
+ *</code>
+ *
+ * In the diagram, all associations are bidirectional;
+ * assocations marked with '*' cascade persist, save, merge operations to the
+ * associated entities (e.g., B cascades persist to D, but D does not cascade
+ * persist to B);
+ *
+ * b, c, d, e, f, and g are all transient unsaved that are associated with each other.
+ *
+ * When persisting b with ORM, the entities are added to the ActionQueue in the following order:
+ * c, d (depends on e), f (depends on d, g), e, b, g.
+ *
+ * Entities are inserted in the following order:
+ * c, e, d, b, g, f.
+ */
+public class CascadeComplicatedTest extends BaseRxTest {
+
+	private B b;
+	private C c;
+	private D d;
+	private E e;
+	private F f;
+	private G g;
+
+	@Override
+	protected Configuration constructConfiguration() {
+		Configuration configuration = super.constructConfiguration();
+		configuration.addAnnotatedClass( B.class );
+		configuration.addAnnotatedClass( C.class );
+		configuration.addAnnotatedClass( D.class );
+		configuration.addAnnotatedClass( E.class );
+		configuration.addAnnotatedClass( F.class );
+		configuration.addAnnotatedClass( G.class );
+		return configuration;
+	}
+
+	@Test
+	public void testPersist(TestContext context) {
+		test(
+				context,
+				openSession()
+						.thenCompose(s -> s.persist(b))
+						.thenCompose(s -> s.flush())
+						.thenCompose(ignore -> check( openSession(), context ))
+		);
+	}
+
+	@Test
+	public void testMergeTransient(TestContext context) {
+		test(
+				context,
+				openSession()
+						.thenCompose(s -> s.merge(b)
+								.thenApply(bMerged -> s)
+						)
+						.thenCompose(s -> s.flush())
+						.thenCompose(v -> check(openSession(), context))
+		);
+	}
+
+	@Test
+	public void testMergeDetached(TestContext context) {
+		test(
+				context,
+				openSession()
+						.thenCompose(s -> s.persist(b))
+						.thenCompose(s -> s.flush())
+						.thenCompose(ignore -> openSession()
+								.thenCompose(s2 -> s2.merge(b))
+						)
+						.thenCompose(v -> check(openSession(), context))
+		);
+	}
+
+
+	@Test
+	public void testRemove(TestContext context) {
+
+		test(
+				context,
+				openSession()
+						.thenCompose(s -> s.persist(b))
+						.thenCompose(s -> s.flush())
+						.thenCompose(ignore -> check( openSession(), context ))
+						.thenAccept(ignore -> {
+							// Cascade-remove is not configured, so remove all associations.
+							// Everything will need to be merged, then deleted in the proper order
+							prepareEntitiesForDelete();
+						})
+						.thenCompose(v -> openSession())
+						.thenCompose(s2 -> s2.merge(b).thenApply(merged -> {
```

Comment:
Merging a detached entity with associations removed results in:
`AssertionFailure( "Unexpected access to the database" )`

This happens when ORM tries to initialize d.e via `SessionImpl#internalLoad`.